pkg/kine/drivers/dqlite/driver.go

```diff
@@ -22,14 +25,37 @@ func init() {
 	}
 }
 
-type Driver = sqlite.Driver
+type Driver struct {
+	*sqlite.Driver
+
+	goDqliteApp *app.App
+}
+
+func (d *Driver) Compact(ctx context.Context, revision int64) (err error) {
+	// Skip the compaction if we're not the leader.
+	//
+	// WARNING: overridden methods will not be visible from the embedded
+	// (sqlite) driver.
+	start := time.Now()
+	isLeader, err := isLocalNodeLeader(ctx, d.goDqliteApp)
+	logrus.Printf(">>> compact isLocalNodeLeader duration : %v", time.Now().Sub(start))
+	if err != nil {
+		logrus.WithError(err).Warning("Couldn't determine whether the local node is the leader, allowing the compaction to proceed")
+	} else if !isLeader {
+		logrus.Info("skipping compaction on follower node")
+		return nil
+	}
+	err = d.Driver.Compact(ctx, revision)
+	logrus.Printf(">>> compact total duration : %v", time.Now().Sub(start))
+	return err
+}
 
 type DriverConfig struct {
 	DB      database.Interface
```

Comment:
I have been wondering in the previous PR whose responsibility was to open the `*sql.DB` interface (i.e. call `sql.Open`).
The answer wasn't clear, so I went for the usual dependency inversion thing. However the addition of an `app.App` field in this struct seems like a good reason to lean in the other direction and call it inside of this method, so that only the app is passed down. 

It also kind of makes sense as the wrapping of a `sql.DB` depends on the system at hand (as the driver is the one who knows how to gain the best performance out of the db, not the caller).

We can do that either here or in another separate PR.

Reply:
I see. ``ConnectionPoolConfig`` would also have to be moved to the dqlite driver, I think.

https://github.com/canonical/k8s-dqlite/blob/26442adc4de6e898ac1c3b34a1ce68817891fa67/pkg/server/server.go#L62-L91

I'd go with a separate PR.

Reply:
Right. Let's do that later on.